File: CPP/ByTier/L/1267_L_Lexicography.cpp
// Problem: CF 1267 L - Lexicography
// https://codeforces.com/contest/1267/problem/L

/*
Purpose: 
This code solves the problem of generating n words of length l each from a given string of n*l letters,
such that the k-th word in lexicographical order is minimized. The algorithm uses sorting and greedy assignment.

Approach:
1. Sort all input characters to get the smallest possible characters available.
2. Construct the first k-1 words by assigning the smallest available characters in order.
3. For the k-th word, we assign characters to minimize it lexicographically while considering that the previous k-1 words are already set.
4. Assign remaining characters to the rest of the words.

Time Complexity: O((n*l)*log(n*l)) due to sorting.
Space Complexity: O(n*l) for storing the words and auxiliary arrays.

Algorithms/Techniques:
- Sorting
- Greedy assignment
- String manipulation
*/

#include <algorithm>
#include <iostream>
#include <string>

using namespace std;
#define ll long long
char a[1005][1005];
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  int n, m, k, tot = 0;
  string s;
  cin >> n >> m >> k >> s;
  sort(s.begin(), s.end()); // Sort all characters to get the lexicographically smallest order
  for (int i = 1; i <= m; i++) {
    for (int j = 0; j < k; j++) {
      // If the current character in previous row matches the k-th word's character,
      // assign the next available character to current position
      if (a[j][i - 1] == a[k - 1][i - 1])
        a[j][i] = s[tot++];
    }
  }
  for (int i = 0; i < n; i++) {
    for (int j = 1; j <= m; j++) {
      // Fill in any remaining positions with the next available characters
      if (a[i][j] == 0)
        a[i][j] = s[tot++];
    }
  }
  for (int i = 0; i < n; i++)
    cout << a[i] + 1 << '\n'; // Print each row (word) starting from index 1
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1267/1267_L_Lexicography.py
# Problem: CF 1267 L - Lexicography
# https://codeforces.com/contest/1267/problem/L

"""
Algorithm/Techniques:
- Count frequency of each character in the input string
- Distribute characters to form n words of length l
- Ensure words are lexicographically ordered
- Minimize the k-th word in lexicographical order

Time Complexity: O(n * l) - We process each character once, and the nested loops are bounded by the input size.
Space Complexity: O(n * l) - To store the answer matrix and character counts.
"""

import sys

input = sys.stdin.buffer.readline


def process(n, l, k, S):
    # Count frequency of each character
    d = [0 for i in range(26)]
    for c in S:
        d[ord(c) - ord("a")] += 1
    
    # Initialize the answer matrix
    answer = [[None for j in range(l)] for i in range(n)]
    
    # Assign first character of first k-1 words to be smallest possible
    I = 0
    for i in range(k):
        while d[I] == 0:
            I += 1
        answer[i][0] = I
        d[I] -= 1
    
    # For the k-th word (to minimize)
    j = 0
    while j + 1 <= l - 1:
        # If this is first word or comparison says we are OK to add more characters
        if k == 1 or (answer[k - 2][j] < answer[k - 1][j]):
            # Fill the k-th word with minimal characters
            for j2 in range(j + 1, l):
                while d[I] == 0:
                    I += 1
                answer[k - 1][j2] = I
                d[I] -= 1
            break
        else:
            # Find the index where to insert new character to maintain lexicographic order
            i1 = k - 1
            while i1 > 0 and answer[i1 - 1][j] == answer[k - 1][j]:
                i1 -= 1
            # Assign remaining positions with minimal characters
            for j2 in range(i1, k):
                while d[I] == 0:
                    I += 1
                answer[j2][j + 1] = I
                d[I] -= 1
            j += 1
    
    # Fill remaining positions with remaining characters
    for i in range(n):
        for j in range(l):
            if answer[i][j] is None:
                while d[I] == 0:
                    I += 1
                answer[i][j] = I
                d[I] -= 1
    
    # Convert integers back to characters and print
    for row in answer:
        row = [chr(x + ord("a")) for x in row]
        row = "".join(row)
        sys.stdout.write(f"{row}\n")


n, l, k = [int(x) for x in input().split()]
S = input().decode().strip()
process(n, l, k, S)


# https://github.com/VaHiX/CodeForces/